### PR TITLE
[WIP] Cross platform foundation

### PR DESCRIPTION
**Pull request by TimUntersberger**
This is a massive refactor that extracts all of the winapi related stuff away from the logic so that we can implement the api for macos and linux later on.

TODO

- [x] combine all of the global static into one STATE struct (WIP)
- [x] display now contains a reference to the focused workspace (WIP)
- [x] display now contains a list of its workspaces (WIP)
- [ ] ~~work_mode is implemented like a user defined mode~~ (BACKLOG)
- [x] move kb_manager into the appstate
- [x] Update custom component documentation
- [x] FIX: Quickly entering/leaving work mode repeatedly causes a deadlock
- [ ] run cargo fmt
- [ ] fix warnings if possible
- [x] run each user provided function in a separate thread

### CODE REVIEW

**Comment by ramirezmike** [comment]
I pulled this down locally and started checking it out, but the first thing I ran into was the same reason the build is failing

>error[E0554]: `#![feature]` may not be used on the stable release channel

It looks like you're wanting to use the shiny, new const_fn. I don't know how you'd go about changing that in the github-actions, but, just curious, if you're definitely sure nog needs const_fn is there a specific version/beta/nightly that you're targeting?

**Comment by TimUntersberger** [comment]
> is there a specific version/beta/nightly that you're targeting

I am using 1.45.0-nightly. I am not really specifically targeting any version, but I think `const_fn` is very ergonomic so I decided to use it.

> I'm seeing some compiler errors I can work through after taking out the nightly part

This is actually compiling for me now locally. I just have to check the functionality. So if you still get some compiliation errors, after using nightly please tell me.

**Comment by ramirezmike** [comment]
I changed to nightly and compiled and it worked then fetched latest and that also worked  👍. I might just start reviewing so I can rebase my pull request on these changes at some point.

**Comment by TimUntersberger** [comment]
@ramirezmike right now the refactor should be almost complete. The only thing is testing whether all of the functionality is still working. If you find any bugs please tell me.

On another note how are you feeling about the API changes? Are you seeing any code smells?

File: src/display.rs [hunk 1]
```diff
@@ -1,173 +1,246 @@
...
-        self.right - self.left
+        self.rect.width()
     }
-    pub fn working_area_height(&self) -> i32 {
-        let task_bar_height = if let Some(task_bar) = self.task_bar {
-            match task_bar.position {
-                task_bar::TaskBarPosition::Top | task_bar::TaskBarPosition::Bottom => {
-                    task_bar.height
+    pub fn is_primary(&self) -> bool {
+        self.rect.left == 0 && self.rect.top == 0
```
Comment (ramirezmike) [inline src/display.rs:33]: this seems like a Windows specific implementation, right? I don't actually know if other OSes handle it differently.

**Comment by ramirezmike** [comment]
> @ramirezmike right now the refactor should be almost complete. The only thing is testing whether all of the functionality is still working. If you find any bugs please tell me.
> 
> On another note how are you feeling about the API changes? Are you seeing any code smells?

I'm still working my way through it but hopefully I'll have more time soon to really test things out.

File: src/bar/component/current_window.rs [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
-            .get_focused_tile()
-            .map(|t| ComponentText::Basic(t.window.title.clone()))
-            .unwrap_or(ComponentText::Basic("".into()))]
-    })
-}
-
 pub fn create() -> Component {
-    Component::new("ActiveMode", Arc::new(render))
+    Component::new(
+        "ActiveMode",
```
Comment (ramirezmike) [inline src/bar/component/current_window.rs:6]: Apologies for being a little sporadic with comments.

I did notice this the other day... this should be "CurrentWindow" here, right?

File: src/bar/component/current_window.rs [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
+        Arc::new(|ctx| {
+            vec![ctx
+                .state
+                .get_current_grid()
```
Comment (ramirezmike) [inline src/bar/component/current_window.rs:10]: Another thing I thought of here, just a comment on changes I think are good. 

the current_window component would appear on every app_bar across all displays but here it's grabbing the window title of the focused grid's focused window, then presumably showing that title on every display as opposed to each display showing its own focused window title.

Which, idk, I think that's definitely debatable on what is preferable, but there could be other scenarios where something display-specific should appear on only that display's appbar.

Correct me if I'm wrong, but it looks like that kind of thing should be easier now with the changes you made to display where it has its grid/focusedgrid/appbar and all that, is that right?

Comment (TimUntersberger) [reply src/bar/component/current_window.rs:10]: ah shit. I didn't want to change the functionality. Thank you!

**Comment by ramirezmike** [comment]
I pulled latest and compiled and noticed two things

1) nog crashes silently when I do my keybinding for toggling view of the app bar
2) The CPU usage is significantly higher than the version I've been using. The following is just after launching nog without doing anything

one

vs on my graphsplit branch

two

**Comment by TimUntersberger** [comment]
@ramirezmike if you don't find anything else then I will merge this into `development` so we can get started with your PR.

**Comment by ramirezmike** [comment]
> @ramirezmike if you don't find anything else then I will merge this into `development` so we can get started with your PR.

I checked out the latest and it resolved both those issues. I played around with it and didn't really find anything else immediately so I'm good with it.

**Pull request merged**